src/platform_specific/affinity.unix.cc

```diff
@@ -1,20 +1,215 @@
+	cpu_set_t available_cores = {};
+	std::unordered_map<thread_type, uint32_t> thread_pinning_plan;
+	std::unordered_map<pthread_t, pinned_thread_state> pinned_threads;
+};
+thread_pinner_state g_state; // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+
+// This helper removes a thread from the pinned_threads map when it ends
+struct thread_remover {
+	thread_remover() = default;
+	thread_remover(const thread_remover&) = delete;
+	thread_remover& operator=(const thread_remover&) = delete;
+	thread_remover(thread_remover&&) = delete;
+	thread_remover& operator=(thread_remover&&) = delete;
+	~thread_remover() {
+		std::lock_guard lock(g_state.mutex);
+		if(g_state.initialized) { g_state.pinned_threads.erase(pthread_self()); }
 	}
+};
+thread_local std::optional<thread_remover> t_remover; // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+
+// Initializes the thread pinning machinery
+// This captures the current thread's affinity mask and sets the thread pinning machinery up
+// Calls to pin_this_thread prior to this call will have no effect
+void initialize(const runtime_configuration& cfg) {
+	std::lock_guard lock(g_state.mutex);
+	assert(!g_state.initialized && "Thread pinning already initialized.");
+	assert(g_state.thread_pinning_plan.empty() && "Thread pinning plan not initially empty.");
+	assert(g_state.pinned_threads.empty() && "Pinned threads not initially empty.");
+
+	g_state.config = cfg;
+
+	const auto ret = sched_getaffinity(0, sizeof(cpu_set_t), &g_state.available_cores);
+	if(ret != 0) {
+		CELERITY_WARN("Error retrieving initial process affinity mask, will disable pinning.");
+		g_state.config.enabled = false;
+		return;
+	}
+
+	// pinned threads per process: user, scheduler, executor, 1 backend worker per device if enabled
+	uint32_t pinned_threads_per_process = 3;
+	if(g_state.config.use_backend_device_submission_threads) { pinned_threads_per_process += g_state.config.num_devices; }
+	// total number of threads to be pinned across processes (legacy mode)
+	const uint32_t total_threads = pinned_threads_per_process * g_state.config.num_legacy_processes;
+
+	if(g_state.config.enabled) {
+		// select the core set to use
+		std::vector<uint32_t> selected_core_ids = {};
+		if(!cfg.hardcoded_core_ids.empty()) {
+			// just use the provided hardcoded IDs
+			if(static_cast<uint32_t>(cfg.hardcoded_core_ids.size()) != total_threads) {
+				CELERITY_WARN("Hardcoded core ID count ({}) does not match the number of threads to be pinned ({}), will disable pinning.",
+				    cfg.hardcoded_core_ids.size(), total_threads);
+			} else {
+				selected_core_ids = cfg.hardcoded_core_ids;
+			}
+		} else {
+			// otherwise, sequential core assignments for now; it is most important that each of the threads is "close"
+			// to the ones next to it in this sequence, so that communication between them is fast
+			selected_core_ids = get_available_sequential_cores(g_state.available_cores, total_threads, cfg.standard_core_start_id);
```

Comment:
```suggestion
			selected_core_ids = get_available_sequential_cores(g_state.available_cores, total_threads, cfg.standard_core_start_id);
			if(selected_core_ids.empty()) {
			  CELERITY_WARN("...")
			}
```

Instead of printing the warning in `get_available_sequential_cores`, we should print it here and include the number of cores that would be required. Otherwise this carries less information than the warning for when pinning is disabled!